Naf query image #126

namirfa opened the pull request:
Fixes Issue #126 

Adds new functionality for Object Recognition Source queries. Queries can now be used to upload locally saved images to VANTIQ. Queries can also be used to delete locally saved images.

The images can be uploaded with user-specified resolution, as described in the README. This can be accomplished by adding a "savedResolution" option to the query parameters ("where" clause in query), exactly as you would for the source configuration.

**TO DO**:
* Need to update the README
* Need to handle the case of duplicate image names (i.e. 2019-02-05--02-35-10.jpg, 2019-02-05--02-35-10(1).jpg, 2019-02-05--02-35-10(2).jpg, etc...)

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionConfigHandler.java:
```diff
@@ -92,10 +92,27 @@ public void handleMessage(ExtensionServiceMessage message) {
+                        source.uploadLocalImages(request, replyAddress);
+                    } else if (operation.equalsIgnoreCase("delete")) {
+                        source.deleteLocalImages(request, replyAddress);
+                    } else {
```
Suppose I send an operation of "nutzo".  I think it will default to something here instead of returning in invalid operation -- which, I think, it should do.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionConfigHandler.java:
```diff
@@ -92,10 +92,27 @@ public void handleMessage(ExtensionServiceMessage message) {
+                        ImageRetrieverResults data = source.retrieveImage(message);
+                        if (data != null) {
+                            source.sendDataFromImage(data, message);
+                        }
```
again, avoid duplication.

```
// get the operation
// if none specified, set operation to the default operation code
//, then do the work based on the operation.

fhcarter commented on objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java:
```diff
@@ -101,32 +89,81 @@ public void saveImage(final BufferedImage image, final String target) {
+                });
+            } catch (IOException e) {
+                LOGGER.error("An error occured while reading the locally saved image file. " + e.getMessage());
+            }
```
Seems like most of this code is duplicated.  Couldn't this do
```
fileToUpload = imgFile
if (queryResize) {
    ... do the resize...
    fileToUpload = tmpFile
} 
// Now do the upload of fileToUpload
```

namirfa replied:
The reason I implemented this with redundant code is because I was getting the following error on line 121 (deleteImage(imgFile) - "Local variable statement defined in an enclosing scope must be final or effectively final."

I can't re-assign the value of imgFile because then the onSuccess callback function cannot delete the temporary file if it needs to. Let me know if that makes sense!

fhcarter replied:
Yeah, don't try & reassign the parameter -- confusing anyway, if call by value vs name vs reference, etc.

IntelliJ seems happy with
```
   public void uploadImage(File imgFile, String target) {
        File fileToUpload = imgFile;
        if (queryResize) {
            try {
                BufferedImage resizedImage = resizeImage(ImageIO.read(imgFile));
                File tmpFile = File.createTempFile("tmp", ".jpg");
                tmpFile.deleteOnExit();
                ImageIO.write(resizedImage, "jpg", tmpFile);
                fileToUpload = tmpFile;

            } catch (IOException e) {
                LOGGER.error("An error occured while reading the locally saved image file. " + e.getMessage());
            }
        } 
        vantiq.upload(fileToUpload,
                "image/jpeg",
                "objectRecognition/" + sourceName + '/'  + target,
                new BaseResponseHandler() {
                    @Override public void onSuccess(Object body, Response response) {
                        super.onSuccess(body, response);
                        LOGGER.trace("Content Location = " + this.getBodyAsJsonObject().get("content"));

                        if (outputDir == null) {
                            deleteImage(imgFile);
                        }
                    }

                    @Override public void onError(List<VantiqError> errors, Response response) {
                        super.onError(errors, response);
                        LOGGER.error("Errors uploading image with VANTIQ SDK: " + errors);
                    }
                });
    }
```

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,362 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+       if (filter == null) {
+           directoryListing = imgDirectory.listFiles();
+       } else {
+           directoryListing = imgDirectory.listFiles(filter);
```
Could this be simplified by setting up a filter that selects everything?

fhcarter commented on objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestUploadAndDeleteQuery.java:
```diff
@@ -0,0 +1,612 @@
+        request.put("imageName", "all");
+        core.uploadLocalImages(request, null);
+                
+        // Checking that all images were uploaded to VANTIQ
```
"all the images are still there?" or aren't there?  Didn't these get inserted in setup()

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,362 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+    * @param request        The parameters sent with the query.
+    * @param replyAddress   The replyAddress used to send a query response.
+    */
+   public void uploadLocalImages(Map<String, ?> request, String replyAddress) {   
```
Here, there's a certain amount of duplicated code plus a bunch of mixed processing.  Seems like there might be a better ways to structure things.  Something like:

// Examine parameters
// Check completeness/legality/etc.
// Compute a filter for appropriate parameter set
// Make call to upload_Whatever_ based on filter (and, possibly, if single)

Issue is that right now, there's a very long routine that's doing a whole bunch of different things intermixed.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,362 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+    * @param request        The parameters sent with the query.
+    * @param replyAddress   The replyAddress used to send a query response.
+    */
+   public void deleteLocalImages(Map<String, ?> request, String replyAddress) {
```
similar comment to uploadLocalImages.

fhcarter commented on objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestUploadAndDeleteQuery.java:
```diff
@@ -0,0 +1,612 @@
+                
+        // Checking that all images were uploaded to VANTIQ
+        Thread.sleep(1000);
+        checkUploadToVantiq(IMAGE_NAME_2);
```
Up to you here, but I find it useful to create a map that has all the related stuff together.  Thus, IMAGE_2 might be a map that has a file name, a date specification, etc., all in the same place.  Easier to track if/when things change later...
Optional, as I said.

fhcarter commented on objectRecognitionSource/src/test/java/io/vantiq/extsrc/objectRecognition/neuralNet/TestUploadAndDeleteQuery.java:
```diff
@@ -0,0 +1,612 @@
+            fail("Image should not have been uploaded to VANTIQ");
+        }
+    }
+}
```
I may have missed it, but I don't see any tests of the actual queries, operations, invalid things, etc.  Seems like we could use the SDK to define procedures & execute them to verify that we got what we expect (or errors when we don't expect).  As these things get more & more complicated, better to have a fuller test system.

namirfa replied:
Almost done with fixes for all the other comments. For this addition to the tests, I was going to add it as part of the fixes for Issue #123 . Let me know if it's alright to add those tests all together, or if I should add them now.

fhcarter commented on objectRecognitionSource/src/main/java/edu/ml/tensorflow/util/ImageUtil.java:
```diff
@@ -101,32 +89,81 @@ public void saveImage(final BufferedImage image, final String target) {
+                            }
+                });
+            } catch (IOException e) {
+                LOGGER.error("An error occured while reading the locally saved image file. " + e.getMessage());
```
Error's not necessarily correct.  Might've been with resizing, etc.

namirfa commented:
Hey @fhcarter 

I took a stab at refactoring the objectRecognitionCore code, but in all honesty I am not sure if I ended up making things even more confusing. If you have time to quickly review that and let me know how I can improve it, that would be much appreciated.

I didn't change the deleteLocalImages method yet, since I'm not 100% confident in the changes I have made to uploadLocalImages().

I added fixes for all of your other comments, and still need to address the README and * duplicate image names I addressed in the initial comment.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,408 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+       List<String> imageDateList = null;
+       ImageUtil imageUtil;
+       
+       // ------------------------ GATHERING PARATMERS ---------------------------
```
Parameters :-)

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,408 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+           imageDir = (String) request.get("imageDir");
+           
+           // Set up ImageUtil properties
+           imageUtil = setupQueryImageUtil(request);
```
Personally, I would probably structure this as "detect negatives then deal with it".  So something like
```
if (!imageDir) {
  send error
} else {
   imageDir = ...
}
imageDate = ...
dateRange = ...
imageName = ...

if (imageName) {
  thingsByName = true
} else {
  if (imageDate & dateRange don't work together) {
    sendError
}
// Now we know things are kosher, deal with them....
```

One thing that might be worth considering:  What if we said that imageDate was always a list.  Then, we short circuit a bunch of this.  If it's size is 2, then we know what's up.  Otherwise, we may have to guess.

that said, what if imageDate was always a list of strings.  Usually, these strings are dates.  But, we also allow "none" or "-" or something to indicate that there's no date.  Then, we turn it into a list of Dates (or date strings or whatever) where the "-" means no date (== -infinity or +infinity).  Then we just use that specification to figure out what we're doing.

Something like that may be less complex to explain -- don't know if it would necessarily be easier to implement, but it probably is because we just check for our list & look for special values.  Things other than "-" (which is probably more international than "none") and valid dates would be right out.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,408 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+               }
+           } else {
+               client.sendQueryError(replyAddress, "io.vantiq.extsrc.objectRecognition.invalidQueryRequest", 
+                       "No imageName or imageDate was specified, or they were incorrectly specified. "
```
Another option to dump all these else's with code in them is to have the various error checkers just set set a code & message (& params if required).  Then, one place at the bottom (or wherever we're sure there's no other error), we make a single call to `client.sendQueryError(replyAddr, code, errorText, errorParams)`. Sometimes that makes things a bit clearer.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,408 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+               uploadOne = false;
+           } else if(dateRange.equals("none")) {
+               uploadOne = true;
+               name = imageDate;
```
Here, if our imageDate is always a list, we create a single 
`createFilter(imageDate)` that just constructs the filter itself.

Or just create a new class
```
class DateRangeFilter implements FileNameFilter {
  beforeDate = null
  afterDate = null
   public DateRangeFilter(List<Date> dateRange) {
      // We assume this list has already been checked for size & dates converted
      beforeDate = dateRange[0] // Either (or both) of these could be null...
      afterDate = dateRange[1]
  }

  @Override
  public boolean accept(File dir, String name) {
        fName =format.parse(name)
			return ((beforeDate == null || fName.after(beforeDate)) &&
                             (afterDate == null || fName.before(afterDate))
		}
    }
}
```
Or something like that (not sure if the logic is quite right, but you get the gist.  The notion being that we'll take the imageDate list & convert it to a list of (2) Dates, either of which could be null.  If both are null, then it's the universal filter.  Otherwise, it does what we need.

The code will look very simple as compared to now, but we'll need a nice explanation of how the filter actually works.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +399,408 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+    */
+   public void uploadOne(String name, String imageDir, ImageUtil imageUtil) {
+       if (!name.endsWith(".jpg")) {
+           name = name + ".jpg";
```
Do we need to deal with non JPEG images?  What if we get a PNG thing in?

namirfa commented:
@fhcarter 

I refactored both the uploadLocalImages() and deleteLocalImages() methods based on your advice. I think it looks much cleaner now, and makes the "imageDate" parameter a bit more straightforward.

Let me know how it looks! If it looks alright, I'll add the check for duplicate image names, and I'll update the README to reflect these changes.

namirfa commented:
@fhcarter 

I added fixes for all of your review comments. Last things to do will still be updating the README and handling duplicate filenames.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +401,302 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+       } else {
+           imageDir = (String) request.get("imageDir");
+           // Set up ImageUtil properties
+           imageUtil = setupQueryImageUtil(request);
```
This is not where I'd expect this to happen, but that's the way it works.  We can add this to the list of restructuring necessary.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +401,302 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+                       dateRange.add(format.parse(date));
+                   } catch (ParseException e) {
+                       client.sendQueryError(replyAddress, "io.vantiq.extsrc.objectRecognition.invalidQueryRequest", 
+                               "One of the dates in the imageDate list could not be parsed.Please be sure that both "
```
need space in "parsed.Please"

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +401,302 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+                       return;
+                   }
+               }
+           }
```
Is there some reason we cannot do all this processing in the handleQueryParameters() call?  That should return, in its map, either an imageName set OR an imageDate that would be a filter.  Then, if imageName is set, we upload/deleteOne (which may no longer be necessary);  otherwise, use the filter.  We may have to send in an "intent" if we need to parameterize the errors to say for "update" vs. "delete".

We have literally the same copied code both places (except for uploadOne vs deleteOne -- both which, if necessary, become "isSingle" or something).  text error compatible.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +401,302 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+       
+       // Calling helper functions to upload images (using filters if needed)
+       
+       if (uploadOne) {
```
I think `uploadOne` goes away -- this becomes `if (imageName != null)`...

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionCore.java:
```diff
@@ -390,6 +401,302 @@ public void sendDataFromImage(ImageRetrieverResults imageResults, ExtensionServi
+               }
+           }
+           deleteOne = false;
+       }
```
more cloned code -- move into `handleQueryParameters()`

namirfa commented:
@fhcarter 

I have updated the code based on your reviewer comments. I also added a check in the DateNameFilter to handle any duplicate filenames. I am editing the README now.

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/ObjectRecognitionConfigHandler.java:
```diff
@@ -212,6 +234,9 @@ boolean createNeuralNet(Map<String, ?> neuralNetConfig ) {
         }
         
+        // Setting the outputDir value for the core, (null if it does not exist)
+        source.outputDir = (String) neuralNetConfig.get(NeuralNetInterface.OUTPUT_DIRECTORY_ENTRY);
```
Minor, but I thought we were going to add the "type" to the interface as well, and use it a few lines above this.

namirfa commented:
@fhcarter 

I have edited the README to reflect the new changes.

fhcarter commented on objectRecognitionSource/README.md:
```diff
@@ -279,6 +279,95 @@ Options available for all Queries (not prepended by anything) are:
     received, instead of being immediately available as a JSON object. This is because Query results are mandated
     to be arrays. Default is false.
+    
+Unique query capabilities available for the YOLO Processor are:
```
... are as follows:

fhcarter commented on objectRecognitionSource/README.md:
```diff
@@ -279,6 +279,95 @@ Options available for all Queries (not prepended by anything) are:
+    
+Unique query capabilities available for the YOLO Processor are:
+
+*   **NOTE:** All YOLO Processor queries must have an "operation" parameter specified. If this parameter is not specified, it 
```
s/must/should/ -- since we say 1) it must be there, and 2) if it's not, we...

So it's a _should_ with a default.

fhcarter commented on objectRecognitionSource/README.md:
```diff
@@ -279,6 +279,95 @@ Options available for all Queries (not prepended by anything) are:
+    that are saved in the output directory, which is defined in the source configuration.
+    *   Parameters:
+        *   "operation": Required. Must be set to "upload".
+        *   *Only one of the following two values *MUST* be specified, or a query error will be returned. If both values are 
```
This section is inconsistent:  first, it says "only one" otherwise an error. Then it says, if both, we'll pick one.  We need to choose between these behaviors.

fhcarter replied:
Perhaps it's trying to say that _at least one_ must be there.  But we're really saying that _exactly one_ should be there.  But if you say both, we'll pick the more specific?

namirfa replied:
I think I will change this to "at least one". Right now, if you specify both then it will pick the more specific one, which is imageName.

fhcarter replied:
We should probably say that we want exactly one.  But say it like "you should specify exactly one of the following..."  then we can give the behavior if you screw up.

fhcarter commented on objectRecognitionSource/README.md:
```diff
@@ -279,6 +279,95 @@ Options available for all Queries (not prepended by anything) are:
+            *   "imageName": A string value representing the name of the file to be uploaded. If set to "all", then all images 
+            in the output directory will be uploaded.
+            *   "imageDate": A list containing two strings, a start date and an end date. All locally saved images falling 
+            between the start and end date, (inclusive), will be uploaded. Dates must be formatted in the following 
```
need to define the "-" date.

Also, would it be better to do "all" as` imageDate: [-,-]`? rather than a special name (that's what we really do anyway :-)

fhcarter replied:
Note that according to the tests, that's how it's done.  We don't seem to have a test for "all" as the imageName.  So maybe we should just remove it.  Use the range of _beginning of time_ to _end of time_ instead.

If we _do not_ make this change, we need to add test for `imageName: "all"`

namirfa replied:
I'm addressing all of your other comments now. We actually do have a test that checks imageName: "all".  It is called `testImageNameUploadAll()`. In any case, I will definitely add the "-" date.

I can remove the "all" behavior entirely if that will simplify things, let me know if you think that's worth doing.

fhcarter replied:
yes, I missed that test.  Sorry.

But I think it might be simpler & clearer to just use the date range of negative-positive infinity instead of "all"

fhcarter commented on objectRecognitionSource/README.md:
````diff
@@ -279,6 +279,95 @@ Options available for all Queries (not prepended by anything) are:
+    	NNsaveImage:"local",
+    	NNoutputDir:"testDir",
+    	NNfileName:"testFile"
+```
````
Need at least one example with open end. Like `imageDate: ["-", "2019-02-08--10-33-36"]`

fhcarter replied:
I think we should add one that shows how to delete or upload everything.

*   Delete Query using imageDate to delete all images:
	```
	SELECT * FROM SOURCE Camera1 AS results WITH
	    	operation:"delete",
	    	imageDate:["-", "-"]
	```

fhcarter commented on objectRecognitionSource/src/main/java/io/vantiq/extsrc/objectRecognition/neuralNet/NeuralNetInterface.java:
```diff
@@ -18,6 +18,7 @@
  * An interface for the neural net that will process the image and return a List of data representing the objects found.
  */
 public interface NeuralNetInterface {
+    static final String OUTPUT_DIRECTORY_ENTRY = "outputDir";
```
need to add (& use) "type" here.

fhcarter commented on objectRecognitionSource/README.md:
````diff
@@ -279,6 +279,95 @@ Options available for all Queries (not prepended by anything) are:
+```
+SELECT * FROM SOURCE Camera1 AS results WITH
+    	operation:"delete",
+    	imageName:"all"
````
note -- we have no tests for this.

fhcarter commented on objectRecognitionSource/README.md:
```diff
@@ -279,6 +279,95 @@ Options available for all Queries (not prepended by anything) are:
+SELECT * FROM SOURCE Camera1 AS results WITH
+    	operation:"upload",
+    	imageName:"2019-02-08--10-33-36.jpg",
+    	savedResolution: {longEdge=600}
```
Does `{longEdge=600}` work?  Doesn't it have to be `{longEdge: 600}`

We need to have tests (maybe part of another test that we need to get in place soon) that test all the examples.  We should include these, or some variation of them, in the sample project.

Working code wins, so we should verify all these queries (albeit with different dates/times)

namirfa replied:
I actually ran this query and had longEdge set with an "=", (which was a mistake), but it did work as expected.

Must be why I didn't notice the swap of "=" for ":".

I will change this in the example to be a colon, but it is interesting that this query works with an equals sign as well.

fhcarter replied:
compiling it down to groovy forgives a multitude of sins :-)

namirfa commented:
@fhcarter 

Added fixes for all your reviewer comments. I did not add the example queries to the sample project yet. There is another issue filed which will add more thorough query tests (using the VANTIQ SDK) for all YOLO queries and No Processor queries. I will add the example queries to the sample project as part of that issue.

fhcarter commented:
Note that this PR is listed as having conflict with ImageUtil.java -- need to merge this in & resolve before (hopefully final) review.

namirfa commented:
@fhcarter 

Resolving that conflict should have been easy, but it all got confused. In any case, it is all resolved now and everything is in place. 

The project builds successfully, and all the tests pass. The README changes are in place as well.

namirfa commented on objectRecognitionSource/README.md:
```diff
@@ -279,6 +279,103 @@ Options available for all Queries (not prepended by anything) are:
+                *   To select *all files* in the output directory, one could use the following value for imageDate: 
+                    *   \["-", "-"\].
+        *   "savedResolution": Optional. This value can be set in the exact same way as it is set in the source configuration. 
+        If it is defined here as a query parameter, it will override the value set in the source configuration.
```
One thing I was thinking of adding to the README, but not sure if it's worth writing out. Say someone set their source configuration to save images locally, with a longEdge of 600. The images would already be resized, and so you would only be able to set a longEdge less than the initial 600 from the source configuration.

Does that make sense? I'm wondering if I should put this in writing somewhere in the README, or if it is already implied since we said earlier that we will not support enlarging images.

fhcarter replied:
I think it makes sense to add.  I think it's hard to explain all the ins & outs, so probably something like "the setting cannot be larger than that provided in the source configuration (since that defines how the images are saved)."  That gets to what's real without having to explain the various machinations that may go on...
